rusty_cards/client/src/p2p_client.rs

```diff
@@ -0,0 +1,213 @@
+        SocketAddr::from_str("127.0.0.1:8080").unwrap()
+    }
+
+    pub fn start(&self) -> io::Result<()> {
+        println!("Looking for an opponent");
+        let opponent_stream: Option<TcpStream> = self.find_opponent();
+        if opponent_stream.is_none() {
+            panic!("Opponent stream is set to None. Unexpected behaviour!");
+        }
+        println!("Ready to start the game :>");
+
+        let mut opponent_stream = opponent_stream.unwrap();
+        if self.going_first.load(Ordering::Relaxed) {
+            println!("Starting the game...");
+            let start_game: Action = Action::Start("Worthy opponent :3".to_string());
+            match send_msg(&mut opponent_stream, start_game) {
+                Ok(()) => println!("Game beggins!"),
+                Err(e) => println!("Connection failed T^T\nError: {e}"),
+            }
+        } else {
+            println!("Waiting for the opponent to start the game...");
+            let mut buffer = [0; 1024];
+            let mut num_bytes = 0;
+            while num_bytes == 0 {
+                num_bytes = opponent_stream.read(&mut buffer).unwrap();
+            }
+
+            let game_beggins: Action = serde_json::from_slice(&buffer[..num_bytes]).unwrap();
+            match game_beggins {
+                Action::Start(greet) => println!("Game beggins! {greet}"),
+                // For now cannot be reached, but later there is a plan to extend Action enum
+                _ => println!("Connection failed T^T"),
+            }
+        }
+        println!("The game ends :] GG");
+        Ok(())
+    }
+
+    // Sends a message to a server with socket address
+    // The response from the server is always one of the follwing:
+    //      1. Wait, along with the password generated by the server
+    //          It indicates that we should wait on a socket that we
+    //          provided for the opponents message that will contain
+    //          the password from the server
+    //      2. Send, along with the opponents socket address and password
+    //          It indicates that we should send a message with a password
+    //          and the opponent will wait for this message on the socket
+    //          address that we have received
+    pub fn find_opponent(&self) -> Option<TcpStream> {
+        let mut server_stream = TcpStream::connect(self.server).unwrap();
+        let connection_established: bool = false;  // this should never change
+        while !connection_established {
+            println!("Sending listener address to the server.");
+            let listener_address = self.listener.local_addr().ok()?;
+            let ready_msg = Handshake::Ready(listener_address);
+            send_msg(&mut server_stream, ready_msg).ok()?;
+
+            println!("Waiting for an opponent...");
+            let mut buffer = [0; 1024];
+            let mut num_bytes = 0;
+            while num_bytes == 0 {
```

Comment:
to bez sensu, jeśli read zwróci zero, to znaczy, że połączenie się przerwało: https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read